### PR TITLE
DAOS plugin for smart_open

### PR DESCRIPTION
**Pull request by sridharbalachandriah**
#### Title

DAOS plugin for smart_open

#### Motivation

Distributed Asynchronous Object Storage (DAOS) is an open source software-defined scale-out object store that provides high bandwidth, low latency, and high I/O operations per second (IOPS) storage containers to HPC applications. This plugin enables python applications using smart_open to connect to DAOS object store through pyDAOS Key-Value interface (https://www.intel.com/content/www/us/en/developer/articles/case-study/unlock-the-power-of-daos-in-python-with-pydaos.html).

If you're adding a new feature, then consider opening a ticket and discussing it with the maintainers before you actually do the hard work.

#### Tests

Unit Tests for DAOS plugin:
1. A generic application with smart_open to open / write / close  and open /read/close to DAOS backend ran successfully.
2. An application with  Ray framework (https://www.ray.io/ , https://github.com/ray-project/ray) using smart_open was written to test the Ray specific smart_open interface. This test also passed with the current pull request with DAOS version 2.2 and 2.3 

#### Work in progress

If you're still working on your PR, include "WIP" in the title.
We'll skip reviewing it for the time being.
Once you're ready to review, remove the "WIP" from the title, and ping one of the maintainers (e.g. mpenkov).

#### Checklist

Before you create the PR, please make sure you have:

- [x] Picked a concise, informative and complete title
- [x] Clearly explained the motivation behind the PR
- [ ] Linked to any existing issues that your PR will be solving
- [x] Included tests for any new functionality
- [x] Checked that all unit tests pass

#### Workflow

Please avoid rebasing and force-pushing to the branch of the PR once a review is in progress.
Rebasing can make your commits look a bit cleaner, but it also makes life more difficult from the reviewer, because they are no longer able to distinguish between code that has already been reviewed, and unreviewed code.

### CODE REVIEW

File: smart_open/daos.py
Comment (cadnce) [inline smart_open/daos.py:28]: This seems non-ideal. As I understand it python3.7 is the minimum supported for smart open https://github.com/RaRe-Technologies/smart_open/pull/688 ?

Comment (sridharbalachandriah) [reply smart_open/daos.py:28]: DAOS supports python with pyDAOS(https://www.intel.com/content/www/us/en/developer/articles/case-study/unlock-the-power-of-daos-in-python-with-pydaos.html) using python3.6. This is packaged separately. Not dependent on the smart_open supported version.

Comment (mpenkov) [reply smart_open/daos.py:24]: Why are you even touching sys.path inside the Python source?

File: smart_open/daos.py
Comment (mpenkov) [inline smart_open/daos.py:3]: @piskvorky We need the copyright transferred to you explicitly, right?

Comment (piskvorky) [reply smart_open/daos.py:3]: Yes. Managing copyright from a multitude of authors would be a nightmare. Too much legal trouble for zero (open source) gain.

Comment (mpenkov) [reply smart_open/daos.py:3]: @sridharbalachandriah Please see the other files for examples of what the copyright header should be.

Comment (sridharbalachandriah) [reply smart_open/daos.py:3]: @mpenkov ,@piskvorky, 

Would the copyright attached below suffice? 

# -*- coding: utf-8 -*-
#
# Copyright (C) 2022 Radim Rehurek <radim@rare-technologies.com>
# Copyright 2022 Hewlett Packard Enterprise Development LP.
#
# This code is distributed under the terms and conditions
# from the MIT License (MIT).
#

Comment (piskvorky) [reply smart_open/daos.py:3]: No, sorry. We want full control over `smart_open`, and having to coordinate with multiple copyright owners complicates that. We're happy to acknowledge your authorship of course, so a header like `Original auhor(s): X Y Z` is perfectly fine.

File: smart_open/daos.py
Comment (sridharbalachandriah) [reply smart_open/daos.py:24]: This is  a requirement when DAOS is not installed from repository packages. Will remove this.

File: smart_open/daos.py
Comment (sridharbalachandriah) [reply smart_open/daos.py:3]: @piskvorky , Thank you for the inputs.

Would the content outlined below work?

**
Copyright (C) 2022 Radim Rehurek [radim@rare-technologies.com](mailto:radim@rare-technologies.com)
Original author(s): X Y Z, Hewlett Packard Enterprise Development LP.

This code is distributed under the terms and conditions
from the MIT License (MIT).**

Comment (piskvorky) [reply smart_open/daos.py:3]: Yes, thanks.

**Review by mpenkov — CHANGES_REQUESTED** [review]
Looks good, I think you also need to add the extra dependencies in setup.py (please see that file for how other backends handle dependencies). Thanks!

**Comment by sridharbalachandriah** [comment]
> Looks good, I think you also need to add the extra dependencies in setup.py (please see that file for how other backends handle dependencies). Thanks!

Thanks, I have added the daos dependency in the setup.py and committed the code.

**Comment by mpenkov** [comment]
Why did you close the PR?

**Comment by sridharbalachandriah** [comment]
> s [ yesterday](#pullrequestreview-1152214593)
I was supposed to close the comment. Not the PR.

File: setup.py
Comment (mpenkov) [inline setup.py:44]: Why did you revert this? The previous commit makes sense to me.

Comment (sridharbalachandriah) [reply setup.py:44]: pydaos as a standalone library is not available. This dependency ends up with errors trying to auto install pydaos. The last set of automated tests failed. pydaos (dependency) gets installed with the DAOS package(rpm).

Comment (mpenkov) [reply setup.py:44]: Ah, I just saw your previous comment. Thank you for explaining this to me (again!).

If pydaos gets installed via an rpm (or otherwise package), then where do the files from the package actually go? I imagine it would be the system's Python, as opposed to the virtual environment. So, that means smart_open would have to be installed into the system's Python as well, right?

Comment (sridharbalachandriah) [reply setup.py:44]: Yes, that's right. By default, pydaos gets installed into system's Python. smart_open installed in the same system's python works fine. 
Example: 
/usr/lib64/python3.6/site-packages/pydaos
/usr/local/lib/python3.6/site-packages/smart_open

If there are different virtual environments created, then one has to link it up appropriately. 
The earlier comment on sys.path.append had the custom path for the pydaos library. pydaos documentation has reference to this. 
sys.path.append("/opt/daos/lib64/python3.6/site-packages/").

Comment (mpenkov) [reply setup.py:44]: @piskvorky What are your thoughts on this? Should we wait for the pydaos thing to be pip-installable before we merge this?

Comment (piskvorky) [reply setup.py:44]: I'll be honest, I don't know what DAOS (besides what's in the PR description), not a user. So my only concern is maintenance and I'm -1 on anything non-standard. We're busy enough as it is, not keen on introducing idiosyncrasies.

Comment (mpenkov) [reply setup.py:44]: @sridharbalachandriah Speaking of tests, how would you test this reliably? I guess locally you have access to the real system, so you can test against that, but what about in CI? Have a look at how some of the other backends are tested: is it possible to **easily** test DAOS in CI?

Comment (sridharbalachandriah) [reply setup.py:44]: @mpenkov , I did have a look at the CI's, a fully functional CI would require to emulate the pydaos(DAOS) functionality. I am checking with the DAOS team for ways to address this.

Comment (sridharbalachandriah) [reply setup.py:44]: @mpenkov , currently DAOS docker container(https://docs.daos.io/v2.2/QSG/docker/) is not published. There is a stand alone mechanism to build DAOS docker container from the rpm, the build has to be  initiated before the DAOS CI tests are invoked. 

Does the smart_open CI infrastructure norms allow for the approach mentioned ?

Comment (mpenkov) [reply setup.py:44]: No, not currently.

I think there are two show-stopper problems for this PR:

1) The underlying library is not pip-installable. Yes, it works under the system Python (some _some_ systems), but using the system Python is generally a bad practice, for several reasons. For example, you cannot expect the system's Python version to be what the user wants. So, it's impossible for people to use this contribution in the general case.
2) It's difficult to test this, so the maintenance burden is also high.

@sridharbalachandriah I really appreciate your effort, but I don't think this contribution is a good fit for smart_open right now.

@piskvorky

**Comment by mpenkov** [comment]
Thank you for your efforts, but this PR is not a good fit for the smart_open project in its current state.

**Pull request closed without merging**
